ibis/backends/clickhouse/compiler/relations.py

```diff
@@ -0,0 +1,199 @@
+    assert not needs_alias, "needs_alias is True"
+    if needs_alias := isinstance(op.table, ops.Join) and not isinstance(
+        op.table, (ops.LeftSemiJoin, ops.LeftAntiJoin)
+    ):
+        args = table.this.args
+        from_ = args["from"]
+        (join,) = args["joins"]
+    else:
+        from_ = join = None
+    tr_val = partial(translate_val, needs_alias=needs_alias, **kw)
+    selections = tuple(map(tr_val, op.selections)) or "*"
+    sel = sg.select(*selections, dialect="clickhouse").from_(
+        from_ if from_ is not None else table, dialect="clickhouse"
+    )
+
+    if join is not None:
+        sel = sel.join(join)
+
+    if predicates := op.predicates:
+        if join is not None:
+            sel = sg.select("*").from_(sel.subquery(kw["aliases"][op.table]))
+        res = functools.reduce(lambda x, y: f"{x} AND {y}", map(tr_val, predicates))
+        sel = sel.where(res, dialect="clickhouse")
+
+    if sort_keys := op.sort_keys:
+        sel = sel.order_by(*map(tr_val, sort_keys), dialect="clickhouse")
+
+    return sel
+
+
+@translate_rel.register(ops.Aggregation)
+def _aggregation(op: ops.Aggregation, *, table, **kw):
+    tr_val = partial(translate_val, **kw)
+
+    by = tuple(map(tr_val, op.by))
+    metrics = tuple(map(tr_val, op.metrics))
+    selections = (by + metrics) or "*"
+    sel = sg.select(*selections).from_(table)
+
+    if by:
+        sel = sel.group_by(*by, dialect="clickhouse")
+
+    if predicates := op.predicates:
+        sel = sel.where(*map(tr_val, predicates), dialect="clickhouse")
+
+    if having := op.having:
+        sel = sel.having(*map(tr_val, having), dialect="clickhouse")
+
+    if sort_keys := op.sort_keys:
+        sel = sel.order_by(*map(tr_val, sort_keys), dialect="clickhouse")
+
+    return sel
+
+
+_JOIN_TYPES = {
+    ops.InnerJoin: "INNER",
+    ops.AnyInnerJoin: "ANY",
+    ops.LeftJoin: "LEFT OUTER",
+    ops.AnyLeftJoin: "LEFT ANY",
+    ops.RightJoin: "RIGHT OUTER",
+    ops.OuterJoin: "FULL OUTER",
+    ops.CrossJoin: "CROSS",
+    ops.LeftSemiJoin: "LEFT SEMI",
+    ops.LeftAntiJoin: "LEFT ANTI",
+}
+
+
+@translate_rel.register
+def _join(op: ops.Join, *, left, right, **kw):
+    predicates = op.predicates
+    if predicates:
+        preds = [sg.condition(translate_val(pred, **kw)) for pred in predicates]
+        on = functools.reduce(lambda left, right: left.and_(right), preds)
+    else:
+        on = None
+    join_type = _JOIN_TYPES[type(op)]
+    try:
+        return left.join(right, join_type=join_type, on=on, dialect="clickhouse")
+    except AttributeError:
+        select_args = (f"{left.alias_or_name}.*",) + (f"{right.alias_or_name}.*",) * (
+            not isinstance(op, (ops.LeftSemiJoin, ops.LeftAntiJoin))
+        )
+        return (
+            sg.select(*select_args, dialect="clickhouse")
+            .from_(left, dialect="clickhouse")
+            .join(right, join_type=join_type, on=on, dialect="clickhouse")
```

Comment:
need to pass clickhouse dialect everywhere?

Reply:
Yeah, unfortunately I haven't figured out a way to avoid this yet. After chatting with Toby, I tried doing this by only returning sqlglot nodes everywhere, but that didn't work. We should do this at some point so I'll add a `TODO`.

Reply:
Done.